### DISCUSSION TITLE
PDB B-factor

### DISCUSSION THREAD

**Discussion by shaharsu** [mdtraj/mdtraj#1870] [Q&A]
Hi,

I'm trying to extract data stored in the B-factor column of a pdb, but can't find any way to do this using mdtraj. Anyone have any ideas?

Thanks!

**Comment by mattwthompson** [comment]
Are you asking about temperature factors? https://www.wwpdb.org/documentation/file-format-content/format33/sect9.html#HETATM

I don't see this anywhere in MDTraj


OpenMM has it somewhere, I'm not sure how to massage it out of this class, though: https://github.com/openmm/openmm/blob/f82876f05d0a3d87927ba32428ee165e056829ab/wrappers/python/openmm/app/internal/pdbstructure.py#L866

**Reply by shaharsu** [reply]
Yes, exactly. This also is where AF2 stores pLDDT values, for example. Too bad this can't be read organically through mdtraj. If anyone is looking here and wants a workaround, you can try this (replace pdb_filename.pdb with your pdb filename):

`B_factor=os.popen('grep \'^ATOM\' pdb_filename.pdb|grep CA|awk \'{print $NF}\'').read().split()`

**Reply by sukritsingh** [reply]
I'm looking at the structure for [PDB files](https://github.com/mdtraj/mdtraj/blob/master/mdtraj/formats/pdb/pdbstructure.py#L986) and I think we may have a way to get them out on a per-atom basis? I think the big "difference" here is that they're called "Temperature factors" in one spot and B-factors in others (ugh), and it's attached to each Atom rather than the PDB file. 
This might be a good issue to raise if there's really no way to do this, given that we can _write_ B-factors out to file!

**Reply by nffaruk** [reply]
Would be a nice feature to have them loaded, they are widely used for extra info, e.g. the openfe package uses them to denote which atoms belong to which end states in an FEP hybrid topology (and interestingly it uses mdtraj to output that hybrid structure with the bfactors...)

**Reply by janash** [reply]
MDTraj does read the temperature factors and occupancies when you give it a PDB file using `md.load`, but this data is not saved to the MDTraj topology that we're used to working with. Internally, MDTraj builds a PdbStructure object, then essentially converts that to an MDTraj Trajectory/Topology to return to th euser. The MDTraj topology does not support all of the information about each atom that is pulled from the PDB, so some of that information is essentially discarded when the topology is created. I know this because I did a [PR recently that added formal charge read from PDBs to the MDTraj topology](https://github.com/mdtraj/mdtraj/pull/1967). It confused me a little at first that you can clearly see temperature factors and occupancies being set for an `Atom` object in `PdbStructure`, and then they seemingly disappear on `Topology`. It's because there are actually two different `Atom` objects ;)

We could consider if we wanted to also pass through the temperature factor and occupancy to the MDTraj topology so that if we use `md.load` for a PDB and then write that structure to a file the data would all be preserved. It would be very similar to the PR that I did for formal charge. The only potential thing to consider is that these attributes would be set to `None` for most topologies, especially ones that did not come from PDBs.

In the meantime, if you really *do* want to extract them, you can use the solution I posed in #1992 (repeated mostly below):

```python
from mdtraj.formats.pdb.pdbstructure import PdbStructure

with open("YOUR_PDB_FILE") as f:
    struct = PdbStructure(f, load_all_models=True)
atoms = list(struct.iter_atoms())
first_atom = atoms[0]

# temperature factor
first_atom.temperature_factor

# occupancy
first_atom.occupancy
```

So, I guess the big question is should we add `temperature_factor` and `occupancy` to `Atom` objects in `Topology`. Somehow it seems to be a common question. :)

**Reply by nffaruk** [reply]
Thanks for the alternative approach in the meantime!

**Reply by sukritsingh** [reply]
> We could consider if we wanted to also pass through the temperature factor and occupancy to the MDTraj topology so that if we use md.load for a PDB and then write that structure to a file the data would all be preserved.

I think given that a) we already read in the temperature factors and occupancies, and b) we can write out the temperature factors, I think it would be great to be able to access that information in the MDTraj topology object. In particular, because this enables users to do end-to-end reading and writing of per-atom statistics (like B-factors and RMSF) which can make it easy to make figures where we map the B-factor to a color-scale per residue in pymol. It seems very in-scope for MDTraj through that lens, at least to me.

> The only potential thing to consider is that these attributes would be set to None for most topologies, especially ones that did not come from PDBs.

I think we could consider either setting them to `0` or `-1` since `-1` is nonsensical for these values anyways for a knowledgable user, but seem easier to handle backend-wise than `None`?

**Reply by mattwthompson** [reply]
I'm weakly -1 on storing these in the core data models provided that they only sensibly come from PDB files and we already have a way of representing them in a PDB-centric data model. I'd like to learn more about use cases that are hindered by using this code path before really taking on the weight (in silicon memory) and possible burden (in carbon-based memory) of dealing with this other use cases that have no clue what these attributes mean.

**Reply by janash** [reply]
I definitely agree that having an understanding of what these are used for is ideal and that it's kind of problematic that we may only expect to have them on a "PDB-centric" model. I am wondering (1) do people want this information on a topology or on a trajectory? (2) what is it used for? 👀 I've seen this question at least twice recently (with the issue referenced). I think people uses these for some analyses, and as @shaharsu mentioned, AlphaFold2 uses these columsn to store pLDDT values.

I'd be totally in favor of simply making the use of `PdbStructure` more prominent for this use case in the documentation if that meets user needs. One thing I'd like to think about is if there is a way to handle information loss when doing something like reading a PDB -> doing some analysis -> writing a PDB. As written, if you read in a PDB with temperature factor or occupany columns, these values will not be preserved when writing the file (again, one could likely direct the user to use `PdbStructure` if this was a concern)
